enhancements/network/on-prem-configurable-lb-dns-stack.md

```diff
@@ -0,0 +1,205 @@
+
+## Release Signoff Checklist
+
+- [X] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+- [ ] User-facing documentation is created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+## Summary
+
+In the current implementation,  the self-hosted DNS/LB stack runs under the MCO umbrella by default in all on-prem platforms and there's no option to configure it. 
+Some customers with high production needs usually have their own external Load Balancing and DNS resolution. In addition, there are cases in which some traffic used to provide these self-hosted services is disallowed in the customer's network.                                                                                                                                                                                                                                        The outcome of such scenarios is that cluster's resources are spent on providing unused services and generating traffic against certain customers network design.
+
+This enhancement suggests a new method for providing the self-hosted stack.
+ 
+*Note:* as per the OCP plan to move API and Ingress to K8S SLB, this direction should be investigated deeper after MetalLb will be integrated.
+
+
+## Motivation
+
+Allow customers to adjust the self hosted DNS/LB stack according to the needs they really have instead of providing them in a generic way. 
+Additionally, detaching the self hosted stack configuration from the selected platform will allow new customers from UPI (platform: None) to leverage the self-hosted stack.
+
+
+### Goals
+
+- Ability to use the self-hosted stack in all on-prem platforms.
+- Ability to control the self-hosted stack configuration.
+- Retrieve the self-hosted stack status through K8S API.
+- Maintain a single code base/manifests for all the on-prem platforms.
+
+
+## Proposal
+
+The [Keepalived and infra DNS components](https://github.com/openshift/machine-config-operator/tree/master/manifests/baremetal) on the bootstrap node will keep running through the MCO.
+However, the services on masters and workers nodes will be managed by a new operator instead of static pods.
+
+The following functionalities are expected as part of the new operator:
+* Introduce a CRD used by the operator to manage the self-hosted stack.
+* Fields in CR spec that control the self-hosted stack configuration.
+* Fields in CR status that expose self-hosted stack status (e.g: which node holds the API-VIP).
+* Run the self-hosted stack components (Kepalived, Coredns-MDNS, HAProxy, MDNS-publisher) as a K8S applications.
+
+## Risks and Mitigations
+
+- Impact on on-prem platforms
+- Upgrades/Downgrades
+
+## Design Details
+
+### Open Questions [optional]
+
+- Is `cluster-hosted-net-services-operator` an acceptable name? 
+
```

Comment:
When the VIP moves to another host, do existing IP connections keep working talking to the old host? I assume this is not the case. This breaks the promise of the Kubernetes API.

In case of GCP and Azure we deploy special iptables rules to let old connection keep working for at least the graceful termination period of up to 135s.

Is anything like that possible or planned with a self-hosted LB ?

Reply:
In other words:

I would like to see a details discussion of how a master node reboot happens, with exact timeline what happens when, i.e. when the VIP is transferred to the new host, and what happens to old and new TCP connections (http/2).

Reply:
These are all very good points,  but since the suggested API loadbalancer will run similarly to the current mode  I would prefer to further analyze the functional aspects of the Load balancer in separate enhancement. 
What do you think?

Reply:
> suggested API loadbalancer will run similarly to the current mode

Aren't we switching exactly this: the mode? Before it was static pod (i.e. it always run, especially during node draining in the same way with the same life-cycle as kube-apiserver), now it should become a deployment or DaemonSet (which?).

Reply:
I meant that we keep the same building blocks and provide HA self-hosted LB for the API 
HAProxy , mdns-publisher and coredns-mDNS should run as Daemonsets